### PR TITLE
Add initial tests for parser

### PR DESCRIPTION
**Pull request by o01eg**
Extracted tests for FOCS parsers from #3309

### CODE REVIEW

File: test/parse/TestScriptParser.cpp [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+    auto game_rules = *Pending::WaitForPendingUnlocked(std::move(game_rules_p));
+    BOOST_REQUIRE(!game_rules.Empty());
+    BOOST_REQUIRE(game_rules.RuleExists("RULE_HABITABLE_SIZE_MEDIUM"));
+    BOOST_REQUIRE(GameRules::Type::TOGGLE == game_rules.GetType("RULE_ENABLE_ALLIED_REPAIR"));
+}
+
+BOOST_AUTO_TEST_CASE(parse_techs) {
+    Pending::Pending<TechManager::TechParseTuple> techs_p = Pending::StartParsing(parse::techs<TechManager::TechParseTuple>, m_scrpiting_dir / "techs");
+    auto techs_tuple = *Pending::WaitForPendingUnlocked(std::move(techs_p));
+    auto techs = std::get<0>(std::move(techs_tuple));
```
Comment (geoffthemedio) [inline test/parse/TestScriptParser.cpp:33]: `auto& [techs, tech_categories, categories_seen] = techs_tuple;`

File: test/parse/TestScriptParser.cpp [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+    Pending::Pending<GameRules> game_rules_p = Pending::StartParsing(parse::game_rules, m_scrpiting_dir / "game_rules.focs.txt");
+    auto game_rules = *Pending::WaitForPendingUnlocked(std::move(game_rules_p));
+    BOOST_REQUIRE(!game_rules.Empty());
+    BOOST_REQUIRE(game_rules.RuleExists("RULE_HABITABLE_SIZE_MEDIUM"));
+    BOOST_REQUIRE(GameRules::Type::TOGGLE == game_rules.GetType("RULE_ENABLE_ALLIED_REPAIR"));
+}
+
+BOOST_AUTO_TEST_CASE(parse_techs) {
+    Pending::Pending<TechManager::TechParseTuple> techs_p = Pending::StartParsing(parse::techs<TechManager::TechParseTuple>, m_scrpiting_dir / "techs");
+    auto techs_tuple = *Pending::WaitForPendingUnlocked(std::move(techs_p));
```
Comment (geoffthemedio) [inline test/parse/TestScriptParser.cpp]: `auto&& [techs, tech_categories, categories_seen] = *Pending::WaitForPendingUnlocked(std::move(techs_p));`

Comment (geoffthemedio) [reply test/parse/TestScriptParser.cpp]: something weird is happening here. for me, the returned tuple has an empty container for index 0, then two invalid references for the other two indicies.

Comment (o01eg) [reply test/parse/TestScriptParser.cpp]: Yes, without this change tests are passed but after it applied I get 

```
error: fatal error: in "TestScriptParser/parse_techs": critical check 1 == tech_categories.count("PRODUCTION_CATEGORY") has failed [1 != 0]
```

Comment (o01eg) [reply test/parse/TestScriptParser.cpp]: It works with

```cpp
auto [techs, tech_categories, categories_seen] = *Pending::WaitForPendingUnlocked(std::move(techs_p));
```

**Comment by geoffthemedio** [comment]
I get
`[2021-03-02 22:55:52.782250] [0x000056b0] [info]    Parsing of "techs" failed with error: boost::filesystem::directory_iterator::construct: The system cannot find the path specified: "C:/Users/g_top/Desktop/FOSDK11/FreeOrionAlt2/default\default-test\scripting\techs"`

**Comment by geoffthemedio** [comment]
probably should put `default-test` within `test` ?

**Comment by geoffthemedio** [comment]
also:
````
 msvc2019/Test/Test.vcxproj         |  3 +++
 msvc2019/Test/Test.vcxproj.filters | 13 +++++++++++--
 2 files changed, 14 insertions(+), 2 deletions(-)

diff --git a/msvc2019/Test/Test.vcxproj b/msvc2019/Test/Test.vcxproj
index fdf369f2c..b9eeba3fb 100644
--- a/msvc2019/Test/Test.vcxproj
+++ b/msvc2019/Test/Test.vcxproj
@@ -170,6 +170,7 @@
   </ItemDefinitionGroup>
   <ItemGroup>
     <ClInclude Include="..\..\test\parse\CommonTest.h" />
+    <ClInclude Include="..\..\test\parse\ParserAppFixture.h" />
     <ClInclude Include="..\..\test\system\ClientAppFixture.h" />
     <ClInclude Include="StdAfx.h" />
   </ItemGroup>
@@ -178,7 +179,9 @@
     <ClCompile Include="..\..\client\ClientNetworking.cpp" />
     <ClCompile Include="..\..\test\parse\CommonTest.cpp" />
     <ClCompile Include="..\..\test\parse\main.cpp" />
+    <ClCompile Include="..\..\test\parse\ParserAppFixture.cpp" />
     <ClCompile Include="..\..\test\parse\TestEnumParser.cpp" />
+    <ClCompile Include="..\..\test\parse\TestScriptParser.cpp" />
     <ClCompile Include="..\..\test\system\ClientAppFixture.cpp" />
     <ClCompile Include="..\..\test\system\SmokeTestGame.cpp" />
     <ClCompile Include="..\..\test\system\SmokeTestHostless.cpp" />
diff --git a/msvc2019/Test/Test.vcxproj.filters b/msvc2019/Test/Test.vcxproj.filters
index 912a4071c..1eee2aab7 100644
--- a/msvc2019/Test/Test.vcxproj.filters
+++ b/msvc2019/Test/Test.vcxproj.filters
@@ -22,10 +22,13 @@
     <ClInclude Include="..\..\test\system\ClientAppFixture.h">
       <Filter>Test\system</Filter>
     </ClInclude>
+    <ClInclude Include="StdAfx.h" />
+    <ClInclude Include="..\..\test\parse\ParserAppFixture.h">
+      <Filter>Test\parse</Filter>
+    </ClInclude>
     <ClInclude Include="..\..\test\parse\CommonTest.h">
       <Filter>Test\parse</Filter>
     </ClInclude>
-    <ClInclude Include="StdAfx.h" />
   </ItemGroup>
   <ItemGroup>
     <ClCompile Include="..\..\test\system\SmokeTestGame.cpp">
@@ -52,9 +55,15 @@
     <ClCompile Include="..\..\test\parse\TestEnumParser.cpp">
       <Filter>Test\parse</Filter>
     </ClCompile>
+    <ClCompile Include="StdAfx.cpp" />
+    <ClCompile Include="..\..\test\parse\TestScriptParser.cpp">
+      <Filter>Test\parse</Filter>
+    </ClCompile>
+    <ClCompile Include="..\..\test\parse\ParserAppFixture.cpp">
+      <Filter>Test\parse</Filter>
+    </ClCompile>
     <ClCompile Include="..\..\test\parse\CommonTest.cpp">
       <Filter>Test\parse</Filter>
     </ClCompile>
-    <ClCompile Include="StdAfx.cpp" />
   </ItemGroup>
 </Project>
\ No newline at end of file
````

**Comment by o01eg** [comment]
I've applied MSVC2019 changes.

About error it looks like you missed `techs` folder in `default-test` because game rules parsed successfully and checks in `ParserAppFixture` successfully passed.

Maybe `GetBinDir` returns relative path which tried to process within `default` folder by `ListDir` function.

**Comment by o01eg** [comment]
Could you start tests again with `--log_level=message` ?

**Comment by geoffthemedio** [comment]
> Could you start tests again with `--log_level=message` ?

````
C:\Users\g_top\Desktop\FOSDK11\FreeOrionAlt2>FreeOrionUnitTests.exe --log_level=message
Running 17 test cases...
Test scripting directory: "default-test\scripting"
C:/Users/g_top/Desktop/FOSDK11/FreeOrionAlt2/test/parse/ParserAppFixture.cpp:13: error: fatal error: in "TestScriptParser/parse_game_rules": critical check m_scripting_dir.is_absolute() has failed
Test scripting directory: "default-test\scripting"
C:/Users/g_top/Desktop/FOSDK11/FreeOrionAlt2/test/parse/ParserAppFixture.cpp:13: error: fatal error: in "TestScriptParser/parse_techs": critical check m_scripting_dir.is_absolute() has failed
````

**Comment by o01eg** [comment]
@geoffthemedio Nice, does working directory contain `default-test` so I can just absolutize this relative path?

**Comment by geoffthemedio** [comment]
> @geoffthemedio Nice, does working directory contain `default-test`

Normally, probably. Could run from anywhere, though...
````
    std::cout << "bindir: " << GetBinDir() << std::endl;
    m_scripting_dir = fs::system_complete(GetBinDir() / "default-test" / "scripting");
    BOOST_TEST_MESSAGE("Test scripting directory: " << m_scripting_dir);
````
somehow outputs:
````
bindir: ""
Test scripting directory: "C:\Users\g_top\Desktop\FOSDK11\FreeOrionAlt2\default-test\scripting"
````
and then the tests run.

**Comment by o01eg** [comment]
Should I use `system_complete`, `weakly_canonical` or `canonical` for this?

**Comment by geoffthemedio** [comment]
The documentation is difficult to interpret, so I'm not sure. `system_complete` seems to work...

But possibly it won't work the same as the main program as written now, though, because you call `InitDirs("")` instead of `InitDirs(argv[0])` so it won't initialize the binary directory the same way as in the clients and server, so the call the `GetBinDir()` just returns `""` instead of the actual binary directory. This results in `GetBinDir() / "default-test" / "scripting"` being a relative path, which is made abolute by appending to the working directory by `system_complete`.

**Comment by o01eg** [comment]
We can get access to command line with https://www.boost.org/doc/libs/1_52_0/libs/test/doc/html/utf/user-guide/test-organization/master-test-suite.html

Could you test again on Windows?

**Comment by geoffthemedio** [comment]
Added some debug output
````
    std::cout << "bindir: " << GetBinDir() << std::endl;
    std::cout << "curpath: " << fs::current_path() << std::endl;
    std::cout << "rel: " << GetBinDir() / "default-test" / "scripting" / ".." / "scripting" << std::endl;
    std::cout << "syscom: " << fs::system_complete(GetBinDir() / "default-test" / "scripting" / ".." / "scripting") << std::endl;
````

and got:
````
C:\Users\g_top\Desktop\FOSDK11\FreeOrionAlt2>FreeOrionUnitTests.exe --log_level=message
Running 17 test cases...
bindir: "C:\Users\g_top\Desktop\FOSDK11\FreeOrionAlt2"
curpath: "C:\Users\g_top\Desktop\FOSDK11\FreeOrionAlt2"
rel: "C:\Users\g_top\Desktop\FOSDK11\FreeOrionAlt2\default-test\scripting\..\scripting"
syscom: "C:\Users\g_top\Desktop\FOSDK11\FreeOrionAlt2\default-test\scripting"
Test scripting directory: "C:\Users\g_top\Desktop\FOSDK11\FreeOrionAlt2\default-test\scripting"
````

**Comment by o01eg** [comment]
Looks like it works on Windows too. Could we merge it?

**Comment by geoffthemedio** [comment]
as above,
> probably should put `default-test` within `test` ?

**Comment by o01eg** [comment]
@geoffthemedio I place `default-test` to the same folder as binary files with `target_dependent_data_symlink_to_build` but not sure how do this on Windows.

**Comment by geoffthemedio** [comment]
I'm not sure if / how that's relevant... Instead of having a `default-test` directory, put the `scripting` directory that contains scripts for testing into `test` and then link to that instead of `default-test`.

On Windows, the build is usully in-source, so you'd use the same relative path as on other OSes.

**Comment by o01eg** [comment]
CMake could have build folder somewhere else so you cannot get path to source tree when run tests or FreeOrion itself.

**Comment by o01eg** [comment]
`target_dependent_data_symlink_to_build` could link only top level folders. 

`target_dependent_data_symlink_to_build(fo_unittest_parse ${PROJECT_SOURCE_DIR}/test/scripting)` generates <build-dir>/scripting. So the only way to have tests works on Windows MSVC project and other OS CMake project with same path is to have top level folder with test resources.

**Comment by geoffthemedio** [comment]
link to `/test/` and then add the `scripting/` in C++

**Comment by o01eg** [comment]
@geoffthemedio CMake already creates `test` folder to store build files and fails with error:

```
CMake Error at /mnt/another/srcs/GIT/freeorion/cmake/TargetDependencies.cmake:124 (message):
  `target_dependent_data_symlink_to_build`: Target path exists and is not a
  symlink.
```

**Comment by geoffthemedio** [comment]
and you can't add a link named scripting in bindir/test/ to sourcedir/test/scripting/ ?

**Comment by o01eg** [comment]
@geoffthemedio Yes, `target_dependent_data_symlink_to_build` can only link into `bindir` folder.

**Comment by geoffthemedio** [comment]
I guess then rename `default-test` to `test-scripting`. I don't want it to have `default` in the name.

Is there anything else that needs to go in what is now `default-test` besides `scripting/`? If not, you could remove the nested directories and just moving the `test/scripting/` directory up to top level and call it `test-scripting`.

**Comment by o01eg** [comment]
No, I suppose other resources don't require separate tests. I moved `default-test/scripting` to `test-scripting`.

**Comment by geoffthemedio** [comment]
seems ok

**Pull request merged**